File: backend/internal/ast/ast.go
// internal/ast/ast.go - AST definitions

package ast

type Node interface{}

type Program struct {
	Statements []Node
}

type VarDecl struct {
	Name  string
	Value Node
}

type Assign struct {
	Name  string
	Value Node
}

type IfStmt struct {
	Condition Node
	Then      []Node
	Else      []Node
}

type WhileStmt struct {
	Condition Node
	Body      []Node
}

type ForStmt struct {
	Var   string
	Start Node
	End   Node
	Body  []Node
}

type FuncDef struct {
	Name   string
	Params []string
	Body   []Node
}

type FuncCall struct {
	Name string
	Args []Node
}

type BinOp struct {
	Op    string
	Left  Node
	Right Node
}

type Literal struct {
	Value float64
}

type StrLit struct {
	Value string
}

type Var struct {
	Name string
}

type ImportStmt struct {
	Name string
}

type Embedded struct {
	Lang string
	Code string
}

type WriteStmt struct {
	Expr Node
}

File: compiler/src/ast.rs
// src/ast.rs - AST definitions

#[derive(Debug, Clone)]
pub enum Ast {
    VarDecl(String, Box<Ast>), // name, value
    Assign(String, Box<Ast>),
    If(Box<Ast>, Vec<Ast>, Option<Vec<Ast>>),
    While(Box<Ast>, Vec<Ast>),
    For(String, Box<Ast>, Box<Ast>, Vec<Ast>), // var, from, to, body
    FuncDef(String, Vec<String>, Vec<Ast>),
    FuncCall(String, Vec<Ast>),
    BinOp(String, Box<Ast>, Box<Ast>),
    Literal(f64),
    StrLit(String),
    Var(String),
    Array(Vec<Ast>),
    Index(String, Box<Ast>), // array name, index
}
